aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -0,0 +1,243 @@
+    context.charge(AGGREGATOR_V2_CREATE_SNAPSHOT_BASE)?;
+
+    match ty_args[0] {
+        Type::U128 => {
+            let input = safely_pop_arg!(args, u128);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u128(
+                input
+            )]))])
+        },
+        Type::U64 => {
+            let input = safely_pop_arg!(args, u64);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u64(
+                input
+            )]))])
+        },
+        _ => {
+            // Check if the type is a string
+            let ty = context
+                .deref()
+                .type_to_fully_annotated_layout(&ty_args[0])?;
+            if let MoveTypeLayout::Struct(MoveStructLayout::WithTypes { type_, .. }) = ty {
+                if type_.name.as_str() == "String"
+                    && type_.module.as_str() == "string"
+                    && type_.address == AccountAddress::ONE
+                {
+                    let input = safely_pop_arg!(args, Struct)
+                        .unpack()?
+                        .next()
+                        .unwrap()
+                        .value_as::<Vec<u8>>()?;
+                    let move_string_value =
+                        Value::struct_(Struct::pack(vec![Value::vector_u8(input)]));
+                    let move_snapshot_value = Value::struct_(Struct::pack(vec![move_string_value]));
+                    return Ok(smallvec![move_snapshot_value]);
+                }
+            }
+            // If not a string, return an error
+            Err(PartialVMError::new(StatusCode::ABORTED)
+                .with_message("Unsupported type supplied to aggregator".to_string())
+                .with_sub_status(0x02_0005)
+                .into())
+        },
+    }
+}
+
+/***************************************************************************************************
+ * native fun copy_snapshot(snapshot: AggregatorSnapshot<Element>): AggregatorSnapshot<Element>;
+ **************************************************************************************************/
+
+fn native_copy_snapshot(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(ty_args.len(), 1);
+    debug_assert_eq!(args.len(), 1);
+    context.charge(AGGREGATOR_V2_COPY_SNAPSHOT_BASE)?;
+
+    match ty_args[0] {
+        Type::U128 => {
+            let value = aggregator_snapshot_u128_info(&safely_pop_arg!(args, StructRef))?;
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u128(
+                value
+            )]))])
+        },
+        Type::U64 => {
+            let value = aggregator_snapshot_u64_info(&safely_pop_arg!(args, StructRef))?;
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u64(
+                value
+            )]))])
+        },
+        _ => {
+            // Check if the type is a string
```

Comment:
I think this part has quite a bit of duplication in all native functions? maybe factor it out?

Reply:
Done.